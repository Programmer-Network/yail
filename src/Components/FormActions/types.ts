import { ButtonVariantEnum } from "../Button/types";

export interface IFormActionsProps {
  onSave?: () => void | Promise<void>;
  onCancel?: () => void;
  onDelete?: () => void | Promise<void>;
  onSaveError?: (error: Error) => void;
  onDeleteError?: (error: Error) => void;
  saveText?: string;
  cancelText?: string;
  deleteText?: string;
  dialogDeleteText?: string;
  dialogDeleteTitle?: string;
  isLoading?: boolean;
  isSaving?: boolean;
  isDeleting?: boolean;
  disabled?: boolean;
  showSave?: boolean;
  showCancel?: boolean;
  showDelete?: boolean;
  saveVariant?: ButtonVariantEnum;
  className?: string;
  size?: "small" | "medium" | "large";
  alignment?: "left" | "center" | "right" | "space-between";
}
