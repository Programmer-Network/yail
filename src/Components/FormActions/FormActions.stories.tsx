import FormActions from ".";

export default {
  title: "FormActions",
  parameters: {
    layout: "centered"
  },
  component: FormActions
};

export const Default = () => {
  const handleSave = () => {
    return new Promise(resolve => {
      setTimeout(() => {
        resolve(true);
      }, 3000);
    });
  };

  const handleCancel = () => {
    console.log("Cancel");
  };

  const handleDelete = async () => {
    return new Promise(resolve => {
      setTimeout(() => {
        resolve(true);
      }, 3000);
    });
  };

  return (
    <div>
      <FormActions
        onSave={handleSave}
        onCancel={handleCancel}
        onDelete={handleDelete}
      />
    </div>
  );
};

export const NoDelete = () => {
  return (
    <div>
      <FormActions onSave={() => {}} onCancel={() => {}} />
    </div>
  );
};
