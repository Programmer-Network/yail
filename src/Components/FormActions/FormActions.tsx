import classNames from "classnames";
import { createRef, useState } from "react";

import Button from "../Button";
import { ButtonVariantEnum } from "../Button/types";
import ConfirmDialog from "../ConfirmDialog";
import { IFormActionsProps } from "./types";

const FormActions = ({
  onSave,
  onCancel,
  onDelete,
  onSaveError,
  onDeleteError,
  saveText = "Save",
  cancelText = "Cancel",
  deleteText = "Delete",
  dialogDeleteText = "This action cannot be undone",
  dialogDeleteTitle = "Confirm Deletion",
  isLoading = false,
  isSaving = false,
  isDeleting = false,
  disabled = false,
  showSave = true,
  showCancel = true,
  showDelete = false,
  saveVariant = ButtonVariantEnum.PRIMARY,
  className = "",
  size = "medium",
  alignment = "left"
}: IFormActionsProps) => {
  const confirmDialogRef = createRef<HTMLDialogElement>();
  const [isDeleteLoading, setIsDeleteLoading] = useState(false);
  const [isSaveLoading, setIsSaveLoading] = useState(false);

  const getSizeClasses = () => {
    switch (size) {
      case "small":
        return {
          gap: "gap-2",
          buttonWidth: "min-w-[80px]"
        };
      case "large":
        return {
          gap: "gap-4",
          buttonWidth: "min-w-[120px]"
        };
      default:
        return {
          gap: "gap-3",
          buttonWidth: "min-w-[100px]"
        };
    }
  };

  const getLayoutClasses = () => {
    const baseClasses = "flex-row";

    switch (alignment) {
      case "center":
        return `flex ${baseClasses} justify-center items-center`;
      case "right":
        return `flex ${baseClasses} justify-end items-center`;
      case "space-between":
        return `flex ${baseClasses} justify-between items-center`;
      default:
        return `flex ${baseClasses} justify-start items-center`;
    }
  };

  const handleSave = async (e: React.MouseEvent) => {
    e.preventDefault();
    e.stopPropagation();

    if (!onSave || disabled || isSaveLoading) {
      return;
    }

    try {
      setIsSaveLoading(true);
      await onSave();
    } catch (error) {
      const errorObj =
        error instanceof Error ? error : new Error(String(error));

      onSaveError?.(errorObj);
    } finally {
      setIsSaveLoading(false);
    }
  };

  const handleCancel = (e: React.MouseEvent) => {
    e.preventDefault();
    e.stopPropagation();

    if (!onCancel || disabled) {
      return;
    }

    onCancel();
  };

  const handleDeleteShow = (e: React.MouseEvent) => {
    e.preventDefault();
    e.stopPropagation();

    if (!confirmDialogRef.current || disabled) {
      return;
    }

    confirmDialogRef.current.showModal();
  };

  const handleDeleteConfirm = async () => {
    if (!onDelete || isDeleteLoading) {
      return;
    }

    try {
      setIsDeleteLoading(true);
      await onDelete();
    } catch (error) {
      const errorObj =
        error instanceof Error ? error : new Error(String(error));

      onDeleteError?.(errorObj);
    } finally {
      setIsDeleteLoading(false);
    }
  };

  const isAnyLoading =
    isLoading || isSaving || isSaveLoading || isDeleting || isDeleteLoading;
  const sizeClasses = getSizeClasses();

  return (
    <div className={classNames("w-full", className)}>
      {showDelete && onDelete && (
        <ConfirmDialog
          ref={confirmDialogRef}
          variant='destructive'
          title={dialogDeleteTitle}
          message={dialogDeleteText}
          confirmText={deleteText}
          onConfirm={handleDeleteConfirm}
          isLoading={isDeleteLoading}
        />
      )}

      <div className={classNames(getLayoutClasses(), sizeClasses.gap)}>
        {showCancel && onCancel && (
          <Button
            variant={ButtonVariantEnum.SECONDARY}
            outlined
            onClick={handleCancel}
            disabled={disabled || isAnyLoading}
            className={sizeClasses.buttonWidth}
            icon={{
              iconName: "Incomplete",
              iconPosition: "left",
              iconClassName: "w-6 h-6"
            }}
          >
            {size !== "small" && cancelText}
          </Button>
        )}

        {/* Delete Button (Secondary/Destructive) */}
        {showDelete && onDelete && (
          <Button
            variant={ButtonVariantEnum.SECONDARY}
            outlined
            onClick={handleDeleteShow}
            disabled={disabled || isAnyLoading}
            icon={{
              iconName: "IconBomb",
              iconPosition: "left",
              iconClassName: "w-6 h-6"
            }}
            className={classNames(
              sizeClasses.buttonWidth,
              "transition-all duration-200 hover:scale-105"
            )}
            aria-label={`${deleteText} - This action cannot be undone`}
          >
            {size !== "small" && deleteText}
          </Button>
        )}

        {/* Save Button (Primary) */}
        {showSave && onSave && (
          <Button
            variant={
              saveVariant === ButtonVariantEnum.SECONDARY
                ? ButtonVariantEnum.SECONDARY
                : ButtonVariantEnum.PRIMARY
            }
            onClick={handleSave}
            disabled={disabled || isAnyLoading}
            isLoading={isSaving || isSaveLoading}
            className={sizeClasses.buttonWidth}
            type='submit'
            icon={{
              iconName: "Complete",
              iconPosition: "left",
              iconClassName: "w-5 h-5"
            }}
          >
            {saveText}
          </Button>
        )}
      </div>
    </div>
  );
};

export default FormActions;
