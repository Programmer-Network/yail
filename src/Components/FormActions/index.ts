export { default } from "./FormActions";
export * from "./types";
