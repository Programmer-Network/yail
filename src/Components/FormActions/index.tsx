import classNames from "classnames";
import { createRef, useState } from "react";

import Button from "Components/Button";
import { ButtonVariantEnum } from "Components/Button/types";
import Dialog from "Components/Dialog";
import Icon from "Components/Icon";
import { H3, Paragraph } from "Components/Typography";

import { IFormActionsProps } from "./types";

const FormActions = ({
  onSave,
  onCancel,
  onDelete,
  saveText = "Save",
  cancelText = "Cancel",
  deleteText = "Delete",
  dialogDeleteText = "This action cannot be undone",
  dialogDeleteTitle = "Confirm Deletion",
  isLoading = false,
  isSaving = false,
  isDeleting = false,
  disabled = false,
  showSave = true,
  showCancel = true,
  showDelete = false,
  saveVariant = "primary",
  className = "",
  size = "medium",
  alignment = "left"
}: IFormActionsProps) => {
  const dialogRef = createRef<HTMLDialogElement>();
  const [isDeleteLoading, setIsDeleteLoading] = useState(false);
  const [isSaveLoading, setIsSaveLoading] = useState(false);

  const getSizeClasses = () => {
    switch (size) {
      case "small":
        return {
          gap: "yl:gap-2",
          buttonWidth: "yl:min-w-[80px]"
        };
      case "large":
        return {
          gap: "yl:gap-4",
          buttonWidth: "yl:min-w-[120px]"
        };
      default:
        return {
          gap: "yl:gap-3",
          buttonWidth: "yl:min-w-[100px]"
        };
    }
  };

  const getLayoutClasses = () => {
    const baseClasses = "yl:flex-row";

    switch (alignment) {
      case "center":
        return `yl:flex ${baseClasses} yl:justify-center yl:items-center`;
      case "right":
        return `yl:flex ${baseClasses} yl:justify-end yl:items-center`;
      case "space-between":
        return `yl:flex ${baseClasses} yl:justify-between yl:items-center`;
      default:
        return `yl:flex ${baseClasses} yl:justify-start yl:items-center`;
    }
  };

  const handleSave = async (e: React.MouseEvent) => {
    e.preventDefault();
    e.stopPropagation();

    if (!onSave || disabled || isSaveLoading) return;

    try {
      setIsSaveLoading(true);
      await onSave();
    } catch (error) {
      console.error("Save operation failed:", error);
    } finally {
      setIsSaveLoading(false);
    }
  };

  const handleCancel = (e: React.MouseEvent) => {
    e.preventDefault();
    e.stopPropagation();

    if (!onCancel || disabled) return;
    onCancel();
  };

  const handleDeleteShow = (e: React.MouseEvent) => {
    e.preventDefault();
    e.stopPropagation();

    if (!dialogRef.current || disabled) return;
    dialogRef.current.showModal();
  };

  const handleDeleteClose = () => {
    if (!dialogRef.current) return;
    dialogRef.current.close();
  };

  const handleDeleteConfirm = async (e: React.MouseEvent) => {
    e.preventDefault();
    e.stopPropagation();

    if (!onDelete || isDeleteLoading) return;

    try {
      setIsDeleteLoading(true);
      await onDelete();
      handleDeleteClose();
    } catch (error) {
      console.error("Delete operation failed:", error);
    } finally {
      setIsDeleteLoading(false);
    }
  };

  const isAnyLoading =
    isLoading || isSaving || isSaveLoading || isDeleting || isDeleteLoading;
  const sizeClasses = getSizeClasses();

  return (
    <div className={classNames("yl:w-full", className)}>
      {/* Enhanced Delete Confirmation Dialog */}
      {showDelete && onDelete && (
        <Dialog
          ref={dialogRef}
          className='yl:p-8 yl:!border-text/20 yl:!border-2 yl:max-w-[480px] yl:rounded-2xl yl:backdrop-blur-sm yl:bg-background/95'
          shouldCloseOnClickOutside={!isDeleteLoading}
        >
          <div className='yl:flex yl:flex-col yl:gap-8'>
            <div className='yl:flex yl:flex-col yl:items-center yl:gap-6'>
              <div className='yl:w-20 yl:h-20 yl:rounded-full yl:bg-gradient-to-br yl:from-text/15 yl:to-text/5 yl:flex yl:items-center yl:justify-center yl:border yl:border-text/20'>
                <Icon
                  iconName='IconBomb'
                  className='yl:w-10 yl:h-10 yl:text-text'
                />
              </div>

              <div className='yl:text-center yl:space-y-3'>
                <H3 className='yl:text-xl yl:font-semibold yl:text-text yl:tracking-tight'>
                  {dialogDeleteTitle}
                </H3>
                <Paragraph className='yl:text-text/70 yl:text-base yl:leading-relaxed yl:max-w-sm'>
                  {dialogDeleteText}
                </Paragraph>
              </div>
            </div>

            <div className='yl:flex yl:gap-3 yl:justify-center'>
              <Button
                variant={ButtonVariantEnum.SECONDARY}
                outlined
                onClick={handleDeleteClose}
                disabled={isDeleteLoading}
                className='yl:w-[100px]'
              >
                Cancel
              </Button>
              <Button
                variant={ButtonVariantEnum.ERROR}
                outlined
                onClick={handleDeleteConfirm}
                isLoading={isDeleteLoading}
                disabled={isDeleteLoading}
                className='yl:w-[100px]'
                icon={{
                  iconName: "IconBomb",
                  iconPosition: "left"
                }}
              >
                {deleteText}
              </Button>
            </div>
          </div>
        </Dialog>
      )}

      {/* Main Action Buttons */}
      <div className={classNames(getLayoutClasses(), sizeClasses.gap)}>
        {/* Primary Actions Group */}
        <div className={classNames("yl:flex", "yl:flex-row", sizeClasses.gap)}>
          {showSave && onSave && (
            <Button
              variant={
                saveVariant === "secondary"
                  ? ButtonVariantEnum.SECONDARY
                  : ButtonVariantEnum.PRIMARY
              }
              onClick={handleSave}
              disabled={disabled || isAnyLoading}
              isLoading={isSaving || isSaveLoading}
              className={sizeClasses.buttonWidth}
              type='submit'
              icon={{
                iconName: "Complete",
                iconPosition: "left",
                iconClassName: "yl:w-5 yl:h-5"
              }}
            >
              {saveText}
            </Button>
          )}

          {showCancel && onCancel && (
            <Button
              variant={ButtonVariantEnum.SECONDARY}
              outlined
              onClick={handleCancel}
              disabled={disabled || isAnyLoading}
              className={sizeClasses.buttonWidth}
              icon={{
                iconName: "Incomplete",
                iconPosition: "left",
                iconClassName: "yl:w-5 yl:h-5"
              }}
            >
              {size !== "small" && cancelText}
            </Button>
          )}
        </div>

        {/* Delete Action */}
        {showDelete && onDelete && (
          <Button
            variant={ButtonVariantEnum.SECONDARY}
            outlined
            onClick={handleDeleteShow}
            disabled={disabled || isAnyLoading}
            icon={{
              iconName: "IconBomb",
              iconPosition: "left",
              iconClassName: "yl:w-6 yl:h-6"
            }}
            className={classNames(
              sizeClasses.buttonWidth,
              "yl:transition-all yl:duration-200 hover:yl:scale-105"
            )}
            aria-label={`${deleteText} - This action cannot be undone`}
          >
            {size !== "small" && deleteText}
          </Button>
        )}
      </div>
    </div>
  );
};

export default FormActions;
