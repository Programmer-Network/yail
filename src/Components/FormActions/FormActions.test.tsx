import { fireEvent, render, screen } from "@testing-library/react";
import { vi } from "vitest";

import FormActions from ".";

describe("FormActions component", () => {
  test("renders correctly - snapshot test", () => {
    const { asFragment } = render(
      <FormActions
        onSave={vi.fn()}
        onCancel={vi.fn()}
        onDelete={vi.fn()}
        showDelete={true}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("renders save and cancel buttons by default", () => {
    render(<FormActions onSave={vi.fn()} onCancel={vi.fn()} />);

    expect(screen.getByText("Save")).toBeInTheDocument();
    expect(screen.getByText("Cancel")).toBeInTheDocument();
  });

  test("renders delete button when showDelete is true", () => {
    render(
      <FormActions
        onSave={vi.fn()}
        onCancel={vi.fn()}
        onDelete={vi.fn()}
        showDelete={true}
      />
    );

    // Look for button with Delete text that has the IconBomb icon
    const deleteButtons = screen.getAllByText("Delete");
    expect(deleteButtons.length).toBeGreaterThanOrEqual(1);

    // Find the button that is part of the main interface (not in a dialog)
    const mainDeleteButton = deleteButtons.find(
      button =>
        button.closest("button")?.getAttribute("type") === "button" &&
        button.closest("[data-testid='mock-confirm-dialog']") === null
    );
    expect(mainDeleteButton).toBeInTheDocument();
  });

  test("does not render delete button when showDelete is false", () => {
    render(
      <FormActions
        onSave={vi.fn()}
        onCancel={vi.fn()}
        onDelete={vi.fn()}
        showDelete={false}
      />
    );

    expect(screen.queryByText("Delete")).not.toBeInTheDocument();
  });

  test("calls onSave function when Save button is clicked", () => {
    const handleSave = vi.fn();
    render(<FormActions onSave={handleSave} onCancel={vi.fn()} />);

    fireEvent.click(screen.getByText("Save"));
    expect(handleSave).toHaveBeenCalledTimes(1);
  });

  test("calls onCancel function when Cancel button is clicked", () => {
    const handleCancel = vi.fn();
    render(<FormActions onSave={vi.fn()} onCancel={handleCancel} />);

    fireEvent.click(screen.getByText("Cancel"));
    expect(handleCancel).toHaveBeenCalledTimes(1);
  });

  test("shows delete confirmation dialog when delete button is clicked", async () => {
    render(
      <FormActions
        onSave={vi.fn()}
        onCancel={vi.fn()}
        onDelete={vi.fn()}
        showDelete={true}
      />
    );

    // Dialog should not be visible initially
    expect(screen.queryByTestId("mock-confirm-dialog")).not.toBeInTheDocument();

    // Find the main delete button and click it
    const deleteButtons = screen.getAllByText("Delete");
    const mainDeleteButton = deleteButtons.find(
      button =>
        button.closest("button")?.getAttribute("type") === "button" &&
        button.closest("[data-testid='mock-confirm-dialog']") === null
    );

    fireEvent.click(mainDeleteButton!.closest("button")!);

    // Check for the ConfirmDialog with destructive variant
    const confirmDialog = screen.getByTestId("mock-confirm-dialog");
    expect(confirmDialog).toBeInTheDocument();
    expect(confirmDialog).toHaveAttribute("data-variant", "destructive");
  });

  test("calls onDelete when confirm button is clicked in delete dialog", async () => {
    const handleDelete = vi.fn();
    render(
      <FormActions
        onSave={vi.fn()}
        onCancel={vi.fn()}
        onDelete={handleDelete}
        showDelete={true}
      />
    );

    // Find the main delete button and click it
    const deleteButtons = screen.getAllByText("Delete");
    const mainDeleteButton = deleteButtons.find(
      button =>
        button.closest("button")?.getAttribute("type") === "button" &&
        button.closest("[data-testid='mock-confirm-dialog']") === null
    );

    fireEvent.click(mainDeleteButton!.closest("button")!);

    // Click the confirm button in the dialog
    const confirmButton = screen.getByTestId("confirm-button");
    fireEvent.click(confirmButton);

    expect(handleDelete).toHaveBeenCalledTimes(1);
  });

  test("renders custom button text", () => {
    render(
      <FormActions
        onSave={vi.fn()}
        onCancel={vi.fn()}
        onDelete={vi.fn()}
        saveText='Update'
        cancelText='Discard'
        deleteText='Remove'
        showDelete={true}
      />
    );

    expect(screen.getByText("Update")).toBeInTheDocument();
    expect(screen.getByText("Discard")).toBeInTheDocument();

    // Find the Remove button in the main interface
    const removeButtons = screen.getAllByText("Remove");
    const mainRemoveButton = removeButtons.find(
      button =>
        button.closest("button")?.getAttribute("type") === "button" &&
        button.closest("[data-testid='mock-confirm-dialog']") === null
    );
    expect(mainRemoveButton).toBeInTheDocument();
  });

  test("disables buttons when disabled prop is true", () => {
    render(<FormActions onSave={vi.fn()} onCancel={vi.fn()} disabled={true} />);

    // Check the actual button elements, not the text content
    const saveButton = screen.getByText("Save").closest("button");
    const cancelButton = screen.getByText("Cancel").closest("button");

    expect(saveButton).toBeDisabled();
    expect(cancelButton).toBeDisabled();
  });

  test("shows loading state on save button when isSaving is true", () => {
    render(<FormActions onSave={vi.fn()} onCancel={vi.fn()} isSaving={true} />);

    const saveButton = screen.getByText("Save").closest("button");
    expect(saveButton).toBeDisabled();
  });

  test("conditionally renders buttons based on show props", () => {
    render(
      <FormActions
        onSave={vi.fn()}
        onCancel={vi.fn()}
        showSave={false}
        showCancel={false}
      />
    );

    expect(screen.queryByText("Save")).not.toBeInTheDocument();
    expect(screen.queryByText("Cancel")).not.toBeInTheDocument();
  });

  test("renders with different size classes", () => {
    const { rerender } = render(
      <FormActions onSave={vi.fn()} onCancel={vi.fn()} size='small' />
    );

    // Look for the main container that has the gap classes
    let container = screen.getByText("Save").closest("div")
      ?.parentElement?.parentElement;
    expect(container).toHaveClass("yl:gap-2");

    rerender(<FormActions onSave={vi.fn()} onCancel={vi.fn()} size='large' />);
    container = screen.getByText("Save").closest("div")
      ?.parentElement?.parentElement;
    expect(container).toHaveClass("yl:gap-4");
  });

  test("renders delete dialog with custom title and message", () => {
    render(
      <FormActions
        onSave={vi.fn()}
        onCancel={vi.fn()}
        onDelete={vi.fn()}
        showDelete={true}
        dialogDeleteTitle='Custom Delete Title'
        dialogDeleteText='Custom delete message'
      />
    );

    // Find the main delete button and click it
    const deleteButtons = screen.getAllByText("Delete");
    const mainDeleteButton = deleteButtons.find(
      button =>
        button.closest("button")?.getAttribute("type") === "button" &&
        button.closest("[data-testid='mock-confirm-dialog']") === null
    );

    fireEvent.click(mainDeleteButton!.closest("button")!);

    expect(screen.getByText("Custom Delete Title")).toBeInTheDocument();
    expect(screen.getByText("Custom delete message")).toBeInTheDocument();
  });

  describe("Error handling", () => {
    test("calls onSaveError when save operation fails", async () => {
      const handleSave = vi.fn().mockRejectedValue(new Error("Save failed"));
      const handleSaveError = vi.fn();

      render(
        <FormActions
          onSave={handleSave}
          onSaveError={handleSaveError}
          onCancel={vi.fn()}
        />
      );

      fireEvent.click(screen.getByText("Save"));

      // Wait for the async operation to complete
      await vi.waitFor(() => {
        expect(handleSaveError).toHaveBeenCalledTimes(1);
        expect(handleSaveError).toHaveBeenCalledWith(
          expect.objectContaining({
            message: "Save failed"
          })
        );
      });
    });

    test("calls onDeleteError when delete operation fails", async () => {
      const handleDelete = vi
        .fn()
        .mockRejectedValue(new Error("Delete failed"));
      const handleDeleteError = vi.fn();

      render(
        <FormActions
          onSave={vi.fn()}
          onCancel={vi.fn()}
          onDelete={handleDelete}
          onDeleteError={handleDeleteError}
          showDelete={true}
        />
      );

      // Find and click the main delete button
      const deleteButtons = screen.getAllByText("Delete");
      const mainDeleteButton = deleteButtons.find(
        button =>
          button.closest("button")?.getAttribute("type") === "button" &&
          button.closest("[data-testid='mock-confirm-dialog']") === null
      );
      fireEvent.click(mainDeleteButton!.closest("button")!);

      // Click the confirm button in the dialog
      const confirmButton = screen.getByTestId("confirm-button");
      fireEvent.click(confirmButton);

      // Wait for the async operation to complete
      await vi.waitFor(() => {
        expect(handleDeleteError).toHaveBeenCalledTimes(1);
        expect(handleDeleteError).toHaveBeenCalledWith(
          expect.objectContaining({
            message: "Delete failed"
          })
        );
      });
    });

    test("converts non-Error objects to Error instances for onSaveError", async () => {
      const handleSave = vi.fn().mockRejectedValue("String error");
      const handleSaveError = vi.fn();

      render(
        <FormActions
          onSave={handleSave}
          onSaveError={handleSaveError}
          onCancel={vi.fn()}
        />
      );

      fireEvent.click(screen.getByText("Save"));

      await vi.waitFor(() => {
        expect(handleSaveError).toHaveBeenCalledTimes(1);
        expect(handleSaveError).toHaveBeenCalledWith(
          expect.objectContaining({
            message: "String error"
          })
        );
      });
    });

    test("converts non-Error objects to Error instances for onDeleteError", async () => {
      const handleDelete = vi
        .fn()
        .mockRejectedValue({ code: 500, message: "Server error" });
      const handleDeleteError = vi.fn();

      render(
        <FormActions
          onSave={vi.fn()}
          onCancel={vi.fn()}
          onDelete={handleDelete}
          onDeleteError={handleDeleteError}
          showDelete={true}
        />
      );

      // Find and click the main delete button
      const deleteButtons = screen.getAllByText("Delete");
      const mainDeleteButton = deleteButtons.find(
        button =>
          button.closest("button")?.getAttribute("type") === "button" &&
          button.closest("[data-testid='mock-confirm-dialog']") === null
      );
      fireEvent.click(mainDeleteButton!.closest("button")!);

      // Click the confirm button in the dialog
      const confirmButton = screen.getByTestId("confirm-button");
      fireEvent.click(confirmButton);

      await vi.waitFor(() => {
        expect(handleDeleteError).toHaveBeenCalledTimes(1);
        expect(handleDeleteError).toHaveBeenCalledWith(
          expect.objectContaining({
            message: "[object Object]"
          })
        );
      });
    });

    test("does not call error handlers when operations succeed", async () => {
      const handleSave = vi.fn().mockResolvedValue(undefined);
      const handleDelete = vi.fn().mockResolvedValue(undefined);
      const handleSaveError = vi.fn();
      const handleDeleteError = vi.fn();

      render(
        <FormActions
          onSave={handleSave}
          onSaveError={handleSaveError}
          onCancel={vi.fn()}
          onDelete={handleDelete}
          onDeleteError={handleDeleteError}
          showDelete={true}
        />
      );

      // Test successful save
      fireEvent.click(screen.getByText("Save"));
      await vi.waitFor(() => {
        expect(handleSave).toHaveBeenCalledTimes(1);
      });

      // Wait for save operation to complete and button to be enabled again
      const saveButton = screen.getByText("Save").closest("button");
      await vi.waitFor(() => {
        expect(saveButton).not.toBeDisabled();
      });

      // Test successful delete
      const deleteButtons = screen.getAllByText("Delete");
      const mainDeleteButton = deleteButtons.find(
        button =>
          button.closest("button")?.getAttribute("type") === "button" &&
          button.closest("[data-testid='mock-confirm-dialog']") === null
      );
      fireEvent.click(mainDeleteButton!.closest("button")!);

      const confirmButton = screen.getByTestId("confirm-button");
      fireEvent.click(confirmButton);

      await vi.waitFor(() => {
        expect(handleDelete).toHaveBeenCalledTimes(1);
      });

      // Error handlers should not have been called
      expect(handleSaveError).not.toHaveBeenCalled();
      expect(handleDeleteError).not.toHaveBeenCalled();
    });

    test("resets loading state after save error", async () => {
      const handleSave = vi.fn().mockRejectedValue(new Error("Save failed"));
      const handleSaveError = vi.fn();

      render(
        <FormActions
          onSave={handleSave}
          onSaveError={handleSaveError}
          onCancel={vi.fn()}
        />
      );

      const saveButton = screen.getByText("Save").closest("button");

      fireEvent.click(screen.getByText("Save"));

      // Button should be disabled during loading
      expect(saveButton).toBeDisabled();

      // Wait for error to be handled
      await vi.waitFor(() => {
        expect(handleSaveError).toHaveBeenCalledTimes(1);
      });

      // Button should be enabled again after error
      expect(saveButton).not.toBeDisabled();
    });

    test("resets loading state after delete error", async () => {
      const handleDelete = vi
        .fn()
        .mockRejectedValue(new Error("Delete failed"));
      const handleDeleteError = vi.fn();

      render(
        <FormActions
          onSave={vi.fn()}
          onCancel={vi.fn()}
          onDelete={handleDelete}
          onDeleteError={handleDeleteError}
          showDelete={true}
        />
      );

      // Find and click the main delete button
      const deleteButtons = screen.getAllByText("Delete");
      const mainDeleteButton = deleteButtons.find(
        button =>
          button.closest("button")?.getAttribute("type") === "button" &&
          button.closest("[data-testid='mock-confirm-dialog']") === null
      );
      fireEvent.click(mainDeleteButton!.closest("button")!);

      // Get the confirm button and check it's not disabled initially
      const confirmButton = screen.getByTestId("confirm-button");
      expect(confirmButton).not.toBeDisabled();

      fireEvent.click(confirmButton);

      // Wait for error to be handled
      await vi.waitFor(() => {
        expect(handleDeleteError).toHaveBeenCalledTimes(1);
      });

      // Confirm button should not be in loading state after error
      expect(confirmButton).not.toHaveAttribute("data-loading", "true");
    });
  });
});
