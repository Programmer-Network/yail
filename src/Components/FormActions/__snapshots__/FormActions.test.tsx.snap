// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FormActions component > renders correctly - snapshot test 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <div
      class="flex flex-row justify-start items-center gap-3"
    >
      <button
        class="select-none px-3 py-2 font-semibold tracking-tight rounded-md border-2 transition-all duration-200 transform hover:scale-105 cursor-pointer border-text/20 text-text/20 fill-text/20 hover:border-text/30 hover:text-text/30 hover:fill-text/30 hover:bg-background/20 min-w-[100px]"
        type="button"
      >
        <div
          class="relative flex items-center justify-center"
        >
          <span
            class="flex items-center gap-1"
          >
            <div
              class="w-6 h-6"
              data-icon="Incomplete"
              data-testid="icon-Incomplete"
            >
              [Icon: Incomplete]
            </div>
            Cancel
          </span>
          <span
            class="absolute"
          />
        </div>
      </button>
      <button
        class="select-none px-3 py-2 font-semibold tracking-tight rounded-md border-2 transition-all duration-200 transform hover:scale-105 cursor-pointer border-text/20 text-text/20 fill-text/20 hover:border-text/30 hover:text-text/30 hover:fill-text/30 hover:bg-background/20 min-w-[100px] transition-all duration-200 hover:scale-105"
        type="button"
      >
        <div
          class="relative flex items-center justify-center"
        >
          <span
            class="flex items-center gap-1"
          >
            <div
              class="w-6 h-6"
              data-icon="IconBomb"
              data-testid="icon-IconBomb"
            >
              [Icon: IconBomb]
            </div>
            Delete
          </span>
          <span
            class="absolute"
          />
        </div>
      </button>
      <button
        class="select-none px-3 py-2 font-semibold tracking-tight rounded-md border-2 transition-all duration-200 transform hover:scale-105 cursor-pointer border-primary bg-primary text-background hover:bg-transparent hover:text-primary min-w-[100px]"
        type="submit"
      >
        <div
          class="relative flex items-center justify-center"
        >
          <span
            class="flex items-center gap-1"
          >
            <div
              class="w-5 h-5"
              data-icon="Complete"
              data-testid="icon-Complete"
            >
              [Icon: Complete]
            </div>
            Save
          </span>
          <span
            class="absolute"
          />
        </div>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
